connection.go

```diff
@@ -616,19 +640,22 @@ func (conn *Connection) getDialTimeout() time.Duration {
 	return dialTimeout
 }
 
-func (conn *Connection) runReconnects() error {
+func (conn *Connection) runReconnects(ctx context.Context) error {
 	dialTimeout := conn.getDialTimeout()
 	var reconnects uint
 	var err error
 
+	t := time.NewTicker(conn.opts.Reconnect)
+	defer t.Stop()
 	for conn.opts.MaxReconnects == 0 || reconnects <= conn.opts.MaxReconnects {
-		now := time.Now()
-
-		ctx, cancel := context.WithTimeout(context.Background(), dialTimeout)
-		err = conn.createConnection(ctx)
+		localCtx, cancel := context.WithTimeout(ctx, dialTimeout)
+		err = conn.connect(localCtx)
 		cancel()
 
 		if err != nil {
+			if ctx.Err() != nil {
+				return err
```

Comment:
If you don't wrap a context error than it would be more correct:
```suggestion
				return ctx.Err()
```

Reply:
This actually breaks the test (`TestConnect_context_cancel`). I suggest leaving the error we got from the dialer since we don't know the error the dialer will return due to the context being cancelled. Or we can change the checked error in the mentioned test.

Reply:
To be honest, this is a very strange construction. Please add a comment then.

Reply:
I agree. Maybe it would be better to change the test. For now I've added a comment.